worker/genericworker/genericworker.go

```diff
@@ -0,0 +1,201 @@
+package genericworker
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"strings"
+
+	"github.com/taskcluster/taskcluster-worker-runner/cfg"
+	"github.com/taskcluster/taskcluster-worker-runner/protocol"
+	"github.com/taskcluster/taskcluster-worker-runner/run"
+	"github.com/taskcluster/taskcluster-worker-runner/worker/worker"
+)
+
+type genericworkerConfig struct {
+	Path       string
+	ConfigPath string
+	// should be []string
+	Args []interface{}
+}
+
+type genericworker struct {
+	runnercfg *cfg.RunnerConfig
+	wicfg     genericworkerConfig
+	cmd       *exec.Cmd
+}
+
+func (d *genericworker) ConfigureRun(state *run.State) error {
+	var err error
+
+	// copy some values from the provisioner metadata, if they are set; if not,
+	// generic-worker will fall back to defaults
+	for cfg, md := range map[string]string{
+		// generic-worker config : providerMetadata
+		// required
+		"publicIp": "public-ipv4",
+		// optional
+		"privateIP":        "local-ipv4",
+		"instanceType":     "instance-type",
+		"instanceID":       "instance-id",
+		"region":           "region",
+		"availabilityZone": "availability-zone",
+		"zone":             "zone",
+	} {
+		v, ok := state.ProviderMetadata[md]
+		if ok {
+			state.WorkerConfig, err = state.WorkerConfig.Set(cfg, v)
+			if err != nil {
+				return err
+			}
+		} else {
+			log.Printf("provider metadata %s not available; not setting config %s", md, cfg)
+		}
+	}
+
+	set := func(key, value string) {
+		var err error
+		// only programming errors can cause this to fail
+		state.WorkerConfig, err = state.WorkerConfig.Set(key, value)
+		if err != nil {
+			panic(err)
+		}
+	}
+
+	// pass all of ProviderMetadata in as workerTypeMetadata
+	state.WorkerConfig, err = state.WorkerConfig.Set("workerTypeMetadata", state.ProviderMetadata)
+	if err != nil {
+		panic(err)
+	}
+
+	workerPoolID := strings.SplitAfterN(state.WorkerPoolID, "/", 2)
```

Comment:
just curious why we need to do this?

Reply:
this is just used to get the workerType and provisionerId. clarified in a comment.